### PR TITLE
Merge master and review branches

### PR DESCRIPTION
**Pull request by ben-farnham**
Hi All,

This branch contains a merge of branches master and the CanModule-rev. This branch can also be merged to the master branch so that we can converge again on the master branch based approach we use in other quasar repo components.

On my environments I've built (in stand alone mode). I haven't been able to test those builds though as I don't have the equipment (but the unit tests pass on both platforms).

==windows==
systec
pkcan
anagate
unitTestMockUpImplementation

==linux==
sockcan
unitTestMockUpImplementation
anagate

Please review and let me know what you think

### CODE REVIEW

**Comment by ben-farnham** [comment]
Reviewing my own stuff now... should probably 'soften' out the implementation specific CMakeLists.txt so that end-users specify where the libs/includes are on their own machine via the cmake invocation command line. Currently this is only possible to do by modifying the CMakeLists.txt (well, unless the locations on your machine just so happen to coincide with the current hardcoded paths.

I'll add something like
```
-DPEAKCAN_LIB_FILE="/some/where/on/disk/"  -DPEAKCAN_INC_DIR="/another/location.lib"
```

File: CanInterface/include/CanBusAccess.h [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+#define CANBUSACCESS_H
+
+#include "CanLibLoader.h"
+#include "CCanAccess.h"
+#include <map>
+#include <string>
+
+#pragma once
+
+using namespace std;
```
Comment (pnikiel) [inline CanInterface/include/CanBusAccess.h]: "using namespace" in an user header file seems wrong practice.

**Comment by ben-farnham** [comment]
OK, done - end users should now be able to build the specific gateway implementations (PEAK, SYSTEC, socketCAN ...) without changing the makefile of each.

Instructions for how to specify the PEAK/SYSTEC/etc paths are in the README and the cmake build also now outputs a fairly comprehensive error message if the gateway resources (headers/includes) are not found.

**Ready for review**

**Comment by pnikiel** [comment]
@ben-farnham honestly impossible for me to review this in realistic time. Too many things altogether, I really don't see precisely the scope of changes. Intermix of Slava's stuff (which itself deserves an independent review) with your CMakeList updates etc...

I propose you move on with the merges and if you like me to review the thing afterwards just let me know and I will do.

File: CanInterface/include/CanBusAccess.h (hunk 1 above)
Comment (ben-farnham) [reply CanInterface/include/CanBusAccess.h]: Yep. This is actually code directly inherited from the CanModule-rev branch, therefore strictly speaking out of the scope of this effort.

But I agree with your comment and, hopefully, trivial. I'll fix it...

Comment (ben-farnham) [reply CanInterface/include/CanBusAccess.h]: done - one line change (deletion)

**Comment by pnikiel** [comment]
yes, I can test it soon. I hope that the test program that I once contributed is still there ;-)

**Comment by meeludwig** [comment]
further to the CMakeLists.txt and the injected build-chains: The CMakeLists.txt should take care of the architecture windows/linux and the source-pull-in or build style depending on the build option chosen. The injected build-chain takes care about machine specifics (where is boost, where is socketcan, where are vendor dll's...) and specifies the build style, by only assigning parameters.

**Comment by meeludwig** [comment]
small thing to improve the CMakeLists even further: include the lines as well
message(STATUS "[${CMAKE_CURRENT_LIST_FILE}:${CMAKE_CURRENT_LIST_LINE}]: message")

**Comment by pnikiel** [comment]
Continuing - yes, now .so are created.
It might be beyond the scope of the review but the following rather elementary code segfaults:

```
    CanLibLoader* loader = CanLibLoader::createInstance("sock");
    loader->openCanBus("vcan0","Unspecified");
```

**Comment by ben-farnham** [comment]
@meeludwig : I like the suggestion to help locate the source of the message. 
```
message(STATUS "[${CMAKE_CURRENT_LIST_FILE}:${CMAKE_CURRENT_LIST_LINE}]: message")
```
I think cmake supports the concept of a macro, that could be the way to go to to help keep our _message()_ calls readable. Also, this is something that could benefit quasar, open62541-compat etc, i.e. outside the CanModule.

I won't add it to this merge - I'm trying as far as possible to constrain the scope of this pull request to a merge between the _master_ and _CanModule-rev_ branches

**Comment by ben-farnham** [comment]
@pnikiel - re: segfault. Thanks, I'll take a look

**Comment by pnikiel** [comment]
Hi @ben-farnham ,

the crash comes from Slava's code:

(gdb) list
220	
221	int CSockCanScan::configureCanBoard(const string name,const string parameters)
222	{
223	    vector<string> parset;
224	    parset = parcerNameAndPar(name,parameters);
225	    m_channelName = parset[1];
226	    return openCanPort();
227	}
228	
229	int CSockCanScan::openCanPort()
(gdb) p parset
$1 = std::vector of length 1, capacity 1 = {"vcan0"}

so there is an access to the vector beyond its actual storage. Such crashes might or not might happen, depending on the crap which is left in the memory.

I propose you open a ticket for this.
Sorry If i confused you. 
Give me 20 more minutes and i'll give you more feedback.
I think i should pass sock:vcan0 rather than vcan0 only.

**Comment by pnikiel** [comment]
OK I managed to open vcan. I noticed that someone made the module unfit to handle vcan:

2018-05-02 16:57.22.213416 [SockCanScan.cpp:297, ERR] sock:vcan0 can_get_state() failed with error code -1, it was not possible to obtain initial port state

In the original version it would just warn but not throw.

So the basic test has passed. There is a lot to be done in the module, if we even look at the warnings compiler produces etc... but that is out of scope. Good luck ;-)

**Comment by ben-farnham** [comment]
Thanks Piotr, your review is much appreciated. 

As you say - still much to do in the module, but not in the scope of the piece of work this pull request encompasses.

Michael's making a final pass over the module too. Unless Michael uncovers anything awful I'll merge this to master by the end of tomorrow.

**Comment by meeludwig** [comment]
here are the small code fixes I see for building it for a standalone non-quasar client, CANX:
========================================================================
CanInterface/src/CanLibExternCWrapper.cpp: 
332 comment out
287 unsigned int
250 unsigned int

CanInterface/src/gettimeofday.cpp:
#include <winsock2.h> is seen by Linux, workaround is to do #ifdef _WIN32 / #endif for the whole file

SockCanScan.cpp: some warnings of unchecked return codes

build chain adaptions
=================
CanInterfaceImplementations/CMakeLists.txt: locate boost and socketcan libs through injected toolchain , so that both libs can be located freely for any build

CanModule/CMakeLists.txt: locate socketcan for linux builds

socketcan/CMakeLists.txt: add lib search path

the buildchain certainly should be cleaned up further, but it works for CANX with these MINOR modifs. I don't see any serious BUILD problem on my side: nice ;-)

**Comment by ben-farnham** [comment]
@meeludwig: No worries, the pull request appears to have survived - the all important 'squash and merge' button is still there so no harm done.

**Comment by meeludwig** [comment]
BTW, CanModuleTest passes on cc7, but I am sure you knew it ;-)

[mludwig@pcbe13632 bin]$ ./CanModuleTest 
[==========] Running 5 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 5 tests from CanModuleTest
[ RUN      ] CanModuleTest.loadWrongLib
StdOutLog::initialize
2018-May-03 13:25:58.823804 [CanLibLoaderLin.cpp:25, DBG] Proceeding to load library LibDoesNotExist
2018-May-03 13:25:58.824455 [CanLibLoaderLin.cpp:37, ERR] Error: could not load dynamic library [libLibDoesNotExist.so], current working directory ["/home/mludwig/CAN/CANX-comp-utils/bin"] error: libLibDoesNotExist.so: cannot open shared object file: No such file or directory
[       OK ] CanModuleTest.loadWrongLib (1 ms)
[ RUN      ] CanModuleTest.loadCorrectLib
2018-May-03 13:25:58.824620 [CanLibLoaderLin.cpp:25, DBG] Proceeding to load library MockUpCanImplementation
[       OK ] CanModuleTest.loadCorrectLib (2 ms)
[ RUN      ] CanModuleTest.openCanBus
2018-May-03 13:25:58.826374 [CanLibLoaderLin.cpp:25, DBG] Proceeding to load library MockUpCanImplementation
2018-May-03 13:25:58.826450 [CanLibLoader.cpp:50, DBG] Creating CCanAccess: FakeName
2018-May-03 13:25:58.826505 [MockCanAccess.cpp:43, INF] createBUS called with name [FakeName] parameters [Unspecified]
2018-May-03 13:25:58.826557 [CanLibLoader.cpp:60, DBG] Adding CCanAccess to the map: FakeName
[       OK ] CanModuleTest.openCanBus (0 ms)
[ RUN      ] CanModuleTest.sendMessage
2018-May-03 13:25:58.826641 [CanLibLoaderLin.cpp:25, DBG] Proceeding to load library MockUpCanImplementation
2018-May-03 13:25:58.826702 [CanLibLoader.cpp:50, DBG] Creating CCanAccess: FakeName9
2018-May-03 13:25:58.826752 [MockCanAccess.cpp:43, INF] createBUS called with name [FakeName9] parameters [Unspecified]
2018-May-03 13:25:58.826801 [CanLibLoader.cpp:60, DBG] Adding CCanAccess to the map: FakeName9
2018-May-03 13:25:58.826851 [MockCanAccess.cpp:56, DBG] sendMessage cobID [1] len [8] message [0x12345678] rtr [0]
2018-May-03 13:25:58.826905 [MockCanAccess.cpp:56, DBG] sendMessage cobID [2] len [8] message [0x22345678] rtr [0]
2018-May-03 13:25:58.826954 [MockCanAccess.cpp:56, DBG] sendMessage cobID [3] len [8] message [0x32345678] rtr [0]
[       OK ] CanModuleTest.sendMessage (1 ms)
[ RUN      ] CanModuleTest.sendRecvEchoMessage
2018-May-03 13:25:58.827025 [CanLibLoaderLin.cpp:25, DBG] Proceeding to load library MockUpCanImplementation
2018-May-03 13:25:58.827080 [CanLibLoader.cpp:50, DBG] Creating CCanAccess: 
2018-May-03 13:25:58.827130 [MockCanAccess.cpp:43, INF] createBUS called with name [] parameters []
2018-May-03 13:25:58.827176 [CanLibLoader.cpp:60, DBG] Adding CCanAccess to the map: 
2018-May-03 13:25:58.827237 [MockCanAccess.cpp:56, DBG] sendMessage cobID [1] len [8] message [0x0000EC40] rtr [0]
2018-May-03 13:25:58.827306 [CanModuleTest.cpp:50, INF] operator() ID [1] len [8] message [0x0000EC40] rtr [0]
[       OK ] CanModuleTest.sendRecvEchoMessage (0 ms)
[----------] 5 tests from CanModuleTest (4 ms total)

[----------] Global test environment tear-down
[==========] 5 tests from 1 test case ran. (4 ms total)
[  PASSED  ] 5 tests.
2018-May-03 13:25:58.827420 [MockCanAccess.cpp:38, DBG] Destroying MockCanAccess object

**Comment by ben-farnham** [comment]
Yes indeed - on windows too. Glad to see you're running them also. The next step is to run them in travis & appveyor CI builds so that every subsequent commit is validated.

**Pull request merged**